README.md
# R package ```ciperm```
Non-parametric confidence intervals using permutation tests

----
This package implements the methodology described in https://arxiv.org/pdf/2111.14966.pdf

The R package is still work in progress. 
The general univariate case (ie. a model and a test that satisifes the conditions of section 2.1 in the paper) can be handled by calling ```ciperm```/```ciperm0``` directly.
Similarly, the general multivariate case in handled by calling ```ciperm```/```ciperm0```. 


```ciperm0``` and ```ciperm0.multi```: The workhorses that perform the actual permutation scheme

```ciperm```: Computes the confidence interval 
(ie. by using quantiles from ```ciperm0```)

```ciperm.multi```: Computes the confidence interval 

```ciperm.twosample```: 
User-friendly function for computing the two-sample confidence interval. 

```ciperm.linreg```: 
User-friendly function for computing confidence interval for the slope in linear regression. 

```alpha.multi```: Computes the joint confidence level (from output of ```ciperm0.multi```)

```adjusted_ci```: Simple bisection algorithm that computes adjusted confidence intervals (from output of ```ciperm0.multi```)

# Installation 
The recommended way is to use the devtools package, e.g. run `devtools::install_github("naolsen/ciperm")` from the R interface.
As the package is only based on R code, no special compilers are needed. 


changes.txt
Changelog

I will do my best to ensure all changes are included here

08-12-21: Univariate confidence intervals, two-sample case
09-12-21: Linear regression
10-12-21: Multivariate confidence regions and joint confidence level. Allow infinities in the univariate permutation tests.
13-12-21: ciperm.twosample and ciperm.linreg now also works with multivariate input. Added ci_adjusted, which adjusts confidence intervals.
17-12-21: Allow infinities in the multivariate case.


R/linreg.R

## Linear model
make.linear.cov.function <- function(x) {
  function(mu) x*mu
}

## Linear regression test statistic
make.linreg.stat <- function(x) {
  function(eps) abs(sum((x - mean(x))*(eps - mean(eps))))
}


#' Confidence interval for the slope parameter in linear regression
#'
#' Calculates the permutation-test based confidence interval for \eqn{\beta_1} in the linear regression model
#' \deqn{y = \beta_0 + \beta_1 x + \epsilon}. Works in both for univariate and multivariate cases (but \code{x} must be univariate).
#'
#' @param x Values of the independent variable. A numeric vector of length N.
#' @param y Values of the dependent variable. A numeric vector of length N or a matrix of dim N x K.
#' @param bounds Bounds forwarded to \code{uniroot} in the search for l and u values.
#' @param ... Additional arguments to \link{ciperm} or \link{ciperm.multi}
#'
#' @return Vector or array of lower and upper ends of the confidence interval
#'
#' @seealso \link{ciperm},  \link{ciperm.multi}
#'
#' @examples
#' x <- runif(20, -1, 1)
#' y <- 3 + 0.8*x + rt(20, df = 4)
#'
#' ciperm.linreg(x, y, bounds = c(-5, 6))
#'
ciperm.linreg <- function(x, y, bounds, ...) {
  ff <- make.linreg.stat(x)
  phif <- make.linear.cov.function(x)

  if (ncol(as.matrix(y)) == 1)
  ciperm(y, phif, statistic = ff, bounds = bounds, ...)

  else ciperm.multi(y, phif, statistic = ff, bounds = bounds, ...)
}

R/twosample.R



## Covariate function for two-sample case
twosample.cov.function <- function(x,y) {
  make.linear.cov.function(c(rep(0, length(x)), rep(1, length(y))))
}


## Test statistic for two-sample case
make.twosample.stat <- function(f) {
  if (!is.factor(f) || length(levels(f)) != 2) stop ("f must be a factor with two levels")

  f <- as.numeric(f)
  g2 <- which(f == 2)
  g1 <- which(f == 1)
  function(x) abs(mean(x[g2]) - mean(x[g1]))
}


#' Confidence interval for the two-sample difference in means
#'
#' Calculates the permutation-test based confidence interval for the difference in means.
#' Works in both for univariate and multivariate cases.
#'
#' @param x Observations from sample 1. A numeric vector of length N_1 or a matrix of dim N_1 x K.
#' @param y Observations from sample 2. A numeric vector of length N_2 or a matrix of dim N_2 x K.
#' @param ... Additional arguments to \link{ciperm} or \link{ciperm.multi}
#'
#' @details It is the confidence interval for mean(y) - mean(x)
#'
#' @return Vector or array of lower and upper ends of the confidence interval
#'
#' @seealso \link{ciperm},  \link{ciperm.multi}
#'
#' @examples
#' x <- rnorm(20)
#' y <- rnorm(15) + 0.3
#'
#' ciperm.twosample(x,y, M = 2000, level = 0.99) ## 99% confidence interval, 2000 permutations.
#'
ciperm.twosample <- function(x, y, ...) {
  ## Set up covariate function and test statistic
  theta <- c(rep(0, NROW(x)), rep(1, NROW(y)))

  f <- factor(theta)
  ff <- make.twosample.stat(f)
  phif <- make.linear.cov.function(theta)

  if (NCOL(y) == 1)
    ciperm(c(x,y), phif, statistic = ff, bounds = c(min(y) - max(x), max(y) - min(x)), ...)

  else ciperm.multi(rbind(x,y), phif, statistic = ff, bounds = c(min(y) - max(x), max(y) - min(x)), ...)

}


R/univariate.R

#' Permutation scheme for non-parametric confidence intervals
#'
#' The actual permutation scheme. Note that \code{estimate} (if provided), \code{statistic} and \code{phi} must match.
#'
#' @param y Data (vector of length N)
#' @param phi Covariate function. Must return a vector of length N
#' @param M Number of permutations
#' @param statistic Test statistic
#' @param bounds Bounds forwarded to \code{uniroot} in the search for l and u values.
#' @param estimate Parameter estimate. Optional
#' @param infinities If TRUE set l_m = -Inf and u_m = Inf, when uniroot fails to find a root.
#' If FALSE, the procedure fails with an error.
#'
#' @return Matrix of dim M x 2
#'
#' @seealso \link{ciperm}
#'
ciperm0 <- function(y, phi, M = 1000, statistic, bounds, estimate, infinities = FALSE) {
  N <- length(y)
  if (length(y) != length(phi(0))) stop("y and covariate lengths differ!")

  lower <- min(bounds)
  upper <- max(bounds)

  ## Missing estimate, optimize and find out
  if (missing(estimate)) {
    estimate <- optimize(function(mu) statistic(y - phi(mu)), interval = bounds)$minimum
    if (estimate < lower + .Machine$double.eps^0.25 || estimate > upper - .Machine$double.eps^0.25 )
      stop("Estimate not in the interior of the interval. Did you specify your function correctly?")
    ## make some check that estimate is sensible..
  }

  ## Permutation scheme
  lus <- matrix(, M, 2)
  for (j in 1:M) {
    s <- sample(N)
    if (infinities)
      lus[j,] <- c(tryCatch(uniroot(function(mu) statistic((y - phi(mu))[s]) - statistic(y - phi(mu)),
                         lower = lower, upper = estimate)$root, error = function(e) -Inf),
                 tryCatch(uniroot(function(mu) statistic((y - phi(mu))[s]) - statistic(y - phi(mu)),
                         lower = estimate, upper = upper)$root, error = function(e) Inf))
    else
      lus[j,] <- c(uniroot(function(mu) statistic((y - phi(mu))[s]) - statistic(y - phi(mu)),
                                  lower = lower, upper = estimate)$root,
                          uniroot(function(mu) statistic((y - phi(mu))[s]) - statistic(y - phi(mu)),
                                  lower = estimate, upper = upper)$root)
  }
  lus
}

#' Confidence interval using permutation test
#'
#' Constructs confidence intervals at the specified level.
#'
#' @param level Confidence level (i.e. 1- \alpha).
#'
#' @details See \link{ciperm0} for additional parameters.
#'
#' @return Vector of lower and upper ends of the confidence interval
#'
#' @seealso \link{ciperm.twosample} for a user-friendly version for the two-sample case.
#' \link{ciperm.linreg} for a user-friendly version for the linear regression case.
#'
ciperm <-  function(y, phi, M = 1000, statistic, bounds, estimate, level = 0.95, infinities = FALSE) {
  lus <- ciperm0(y, phi, M = M, statistic = statistic, bounds = bounds, estimate = estimate, infinities = infinities)

  out <- c(quantile(lus[,1], probs = 1-level), quantile(lus[,2], probs = level))
  if (infinities && (out[1] == -Inf || out[2] == Inf))
    warning("Confidence interval contains infinity. Did you make your bounds wide enough?")
  names(out) <- c("lower", "upper")
  out
}





R/multivariate.R

#' Permutation scheme for non-parametric confidence intervals (multivariate version)
#'
#' The actual permutation scheme, and the multivariate version of \link{ciperm0}.
#'
#' @param y Matrix of dimension N x K with at least two columns.
#' @param estimate Parameter estimate (vector of length K). Optional
#'
#' @details See \link{ciperm0} for additional parameters.
#' \code{phi}, \code{statistic} and \code{bounds} are assumed univariate and similar across coordinates
#'
#' @return array of dim M x K x 2
#' @export
#'
ciperm0.multi <- function(y, phi, M = 1000, statistic, bounds, estimate,  infinities = FALSE) {

  if (!is.matrix(y) || ncol(y) == 1) stop("Data is of incorrect format")

  N <- nrow(y)
  if (N != length(phi(0))) stop("y and covariate lengths differ!")
  K <- ncol(y)

  lower <- min(bounds)
  upper <- max(bounds)

  ## Missing estimate, optimize and find out
  if (missing(estimate)) {
  estimate <- numeric(K)
  for (k in 1:K) {
    estimate[k] <- optimize(function(mu) statistic(y[,k] - phi(mu)), interval = bounds)$minimum
    if (estimate[k] < lower + .Machine$double.eps^0.25 || estimate[k] > upper - .Machine$double.eps^0.25 )
      stop("Estimate not in the interior of the interval. Did you specify your function correctly?")
    ## make some check that estimate is sensible..
    }
  }

  ## Permutation scheme
  lusm <- array(, c(M, K, 2))
  for (j in 1:M) {
    s <- sample(N)
    for (k in 1:K) {
      if (infinities)
        lusm[j,k,] <- c(tryCatch(uniroot(function(mu) statistic((y[,k] - phi(mu))[s]) - statistic(y[,k] - phi(mu)),
                                      lower = lower, upper = estimate[k])$root, error = function(e) -Inf),
                     tryCatch(uniroot(function(mu) statistic((y[,k] - phi(mu))[s]) - statistic(y[,k] - phi(mu)),
                                      lower = estimate[k], upper = upper)$root, error = function(e) Inf))

      else
        lusm[j,k,] <- c(uniroot(function(mu) statistic((y[,k] - phi(mu))[s]) - statistic(y[,k] - phi(mu)),
                              lower = lower, upper = estimate[k])$root,
                      uniroot(function(mu) statistic((y[,k] - phi(mu))[s]) - statistic(y[,k] - phi(mu)),
                              lower = estimate[k], upper = upper)$root)
    }
  }
  lusm
}


#' Multivariate confidence intervals using permutation tests
#'
#' Constructs confidence intervals at the specified level, and optionally calculates the joint confidence level.
#'
#' @param level Coverage level (i.e. 1- \eqn{\alpha}).
#' @param multilevel Calculate the joint coverage level using \link{alpha.multi}?
#'
#' @details See \link{ciperm0.multi} for additional parameters.
#'
#' @return A 2 x K array of K confidence intervals.
#' @export
#'
#' @seealso \link{ciperm}
#'
ciperm.multi <- function(y, phi, M = 1000, statistic, bounds, estimate, level = 0.95,
                         multilevel = TRUE, infinities = FALSE) {
  lus <- ciperm0.multi(y, phi, M = M, statistic = statistic, bounds = bounds, estimate = estimate, infinities = infinities)
  dl <- dim(lus)
  K <- dl[2]

  out <- matrix(, 2, K)

  for (j in 1:K) {
    out[,j] <- c(quantile(lus[,j,1], probs = 1-level), quantile(lus[,j,2], probs = level))
  }
  rownames(out) <- c("lower", "upper")
  if (infinities && any(is.infinite(out)))
    warning("Confidence interval contains infinity. Did you make your bounds wide enough?")
  if (multilevel) attr(out, "alpha_multiple") <- alpha.multi(lus[,,1], lus[,,2], 1-level)
  out
}


#' Joint coverage level
#'
#' @param lows matrix of l values
#' @param ups matrix of u values
#' @param alpha Significance level (for unadjusted/single confidence intervals)
#'
#' @return Calculates alpha_multiple from the output of \link{ciperm0.multi}
#' @export
#'
alpha.multi <- function(lows, ups, alpha) {

  stopifnot(dim(lows) == dim(ups))
  K <- ncol(lows)
  M <- nrow(lows)

    c0 <- 0:(2^K-1)
    corners <- matrix(, 2^K, 0)
    for (i in 1:K) corners <- cbind(corners, c0 %% 2^i %/% 2^(i-1))

    dds <- numeric(2^K)
    for (j in 1:2^K) {
      l <- corners[j,]

      dd <- rep(F, M)
      for(u in 1:K) {
        if (l[u] == 0) dd <- dd | lows[,u] < quantile(lows[,u], alpha)
        else dd <- dd | ups[,u] > quantile(ups[,u], 1-alpha)
      }
      dds[j] <-  sum(dd)
    }
    dds
    max(dds)/M
}


#' Adjusted (multivariate) confidence intervals
#'
#' Adjust coverage level such that joint coverage level corresponds
#'
#' @param level
#' @param tol tolerance for reaching \code{level}.
#' @param include_unadjusted Include unadjusted confidence intervals as an attribute?
#'
#' @details This function uses output from \code{ciperm0.multi}.
#' In a simple bisection algorithm, the confidence level is adjusted until the joint coverage level (form \code{alpha.multi}) is \code{code}.
#'
#' @return Adjusted confidence intervals with the adjusted confidence level as an attribute.
#' @export
#'
#' @seealo \link{alpha.multi}
adjusted_ci <- function(lus, level = 0.95, tol = 0.001, include_unadjusted = TRUE) {

  K <- dim(lus)[2]

  out <- matrix(, 2, K)

  for (j in 1:K) {
    out[,j] <- c(quantile(lus[,j,1], probs = 1-level), quantile(lus[,j,2], probs = level))
  }
  rownames(out) <- c("lower", "upper")

  a_target <- 1-level
  a_max <- a_target
  a_min <- 0
  a <- alpha.multi(lus[,,1], lus[,,2], 1-level)
  a_new <- a_max

  while(abs(a -a_target) > tol) {

    if (a_max - a_min < 1e-7) stop("Convergence error!")

    a_new <- (a_min + a_max) / 2
    a <- alpha.multi(lus[,,1], lus[,,2], a_new)

    if (a < a_target) a_min <- a_new
    else a_max <- a_new
  }

  out2 <- matrix(, 2, K)

  for (j in 1:K) {
    out2[,j] <- c(quantile(lus[,j,1], probs = a_new), quantile(lus[,j,2], probs = 1-a_new))
  }
  rownames(out2) <- c("lower", "upper")

  attr(out2, 'alpha_adjusted') <- a_new

  if (include_unadjusted) attr(out2, 'unadjusted.ci') <- out
  out2
}
